mujoco_warp/_src/collision_convex.py

```diff
@@ -967,34 +970,66 @@ def convex_narrowphase(m: Model, d: Data):
   kernel for each type of convex collision pair present in the model, avoiding unnecessary
   computations for non-existent pair types.
   """
-  if not any(m.geom_pair_type_count[upper_trid_index(len(GeomType), g[0].value, g[1].value)] for g in _CONVEX_COLLISION_PAIRS):
+
+  def _pair_count(p1: int, p2: int) -> int:
+    return m.geom_pair_type_count[upper_trid_index(len(GeomType), p1, p2)]
+
+  # no convex collisions, early return
+  if not any(_pair_count(g[0].value, g[1].value) for g in _CONVEX_COLLISION_PAIRS):
     return
 
+  epa_iterations = m.opt.ccd_iterations
+
+  # count of smooth-smooth geom collisions
+  c1 = _pair_count(GeomType.ELLIPSOID.value, GeomType.ELLIPSOID.value)
+  c2 = _pair_count(GeomType.ELLIPSOID.value, GeomType.CYLINDER.value)
+  c3 = _pair_count(GeomType.CYLINDER.value, GeomType.CYLINDER.value)
+  nsmooth = c1 + c2 + c3
+  nsemismooth = 0
+
+  # count of smooth-discrete geom collisions
+  if not nsmooth:
+    s1 = _pair_count(GeomType.HFIELD.value, GeomType.ELLIPSOID.value)
+    s2 = _pair_count(GeomType.HFIELD.value, GeomType.CYLINDER.value)
+    s3 = _pair_count(GeomType.SPHERE.value, GeomType.ELLIPSOID.value)
+    s4 = _pair_count(GeomType.CAPSULE.value, GeomType.ELLIPSOID.value)
+    s5 = _pair_count(GeomType.CAPSULE.value, GeomType.CYLINDER.value)
+    s6 = _pair_count(GeomType.ELLIPSOID.value, GeomType.BOX.value)
+    s7 = _pair_count(GeomType.ELLIPSOID.value, GeomType.MESH.value)
+    s8 = _pair_count(GeomType.CYLINDER.value, GeomType.BOX.value)
+    s9 = _pair_count(GeomType.CYLINDER.value, GeomType.MESH.value)
+    nsemismooth = s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8 + s9
+
```

Comment:
for mesh<>mesh is it possible to determine the maximum number of iterations required?